### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by whatisinternet** [whatisinternet/inflector#18]
Iterating over every char is probably not the best approach. Should probably come up with something better.

**Comment by calebmer** [comment]
I started my own implementation of some of the cases for this library which removes some of the semantic language-dependent assumptions. If you'd like to see my implementation look [here](https://github.com/calebmer/inflections/blob/master/src/case.rs).

I don't think there is a good way to do this besides iterating over every `char` and using functions to get `char` positions likely internally use an iterator.

**Comment by whatisinternet** [comment]
Yeah agreed. The implementation you have is much cleaner. I think that going down a similar path is probably the correct way forward. :smile:

**Comment by NullVoxPopuli** [comment]
speaking of performance issues, https://www.reddit.com/r/ruby/comments/53wec6/how_do_i_should_i_be_getting_performance_boosts/

basically:
```
$ ruby benchmark.rb 

Comparison:
         Ruby: camel:    13179.9 i/s
         Rust: camel:        1.9 i/s - 6978.60x  slower


Comparison:
   Rust: camel_lower:   101904.9 i/s
   Ruby: camel_lower:    10441.9 i/s - 9.76x  slower


Comparison:
          Rust: dash:   230069.3 i/s
          Ruby: dash:    24660.9 i/s - 9.33x  slower


Comparison:
     Ruby: unaltered:  6053346.9 i/s
     Rust: unaltered:  2143459.5 i/s - 2.82x  slower


Comparison:
    Rust: underscore:   271433.8 i/s
    Ruby: underscore:   166316.0 i/s - 1.63x  slower

```

:-\

**Comment by NullVoxPopuli** [comment]
@calebmer I swapped out @whatisinternet's case transforms with yours, and got the following benchmarks (same test as above)

```
$ ruby benchmark.rb 

Comparison:
         Rust: camel:   238774.9 i/s
         Ruby: camel:    13107.3 i/s - 18.22x  slower


Comparison:
   Rust: camel_lower:   230641.5 i/s
   Ruby: camel_lower:     8244.1 i/s - 27.98x  slower


Comparison:
          Rust: dash:   243197.8 i/s
          Ruby: dash:    20350.9 i/s - 11.95x  slower


Comparison:
     Ruby: unaltered:  6030498.8 i/s
     Rust: unaltered:  1960989.6 i/s - 3.08x  slower


Comparison:
    Rust: underscore:   291052.3 i/s
    Ruby: underscore:   169458.2 i/s - 1.72x  slower

```

**Comment by whatisinternet** [comment]
Hey! Thanks for the benchmarks. :+1: I'm aware that the current implementation isn't perfect and I haven't given it enough focus, sorry. That said, I'm very open to people helping and submitting PRs if you're interested :smile:.

@NullVoxPopuli

**Comment by whatisinternet** [comment]
Fixed via #26
